core/vm/exit_code/exit_code.hpp

```diff
@@ -111,6 +143,170 @@ OUTCOME_HPP_DECLARE_ERROR(fc::vm, VMFatal);
 OUTCOME_HPP_DECLARE_ERROR(fc::vm, VMAbortExitCode);
 
 namespace fc::vm {
+  /**
+   * Aborts execution if res has error
+   * @tparam T - result type
+   * @param res - result to check
+   * @param default_error - default VMExitCode to abort with.
+   * @return If res has no error, success() returned. Otherwise if res.error()
+   * is VMAbortExitCode or VMFatal, the res.error() returned, else
+   * if res.error()is VMExitCode, the aborted res.error()returned, else
+   * aborted default_error returned
+   */
+  template <typename T>
+  outcome::result<void> requireNoError(const outcome::result<T> &res,
+                                       const VMExitCode &default_error) {
+    if (res.has_error()) {
+      if (isFatal(res.error()) || isAbortExitCode(res.error())) {
+        return res.error();
+      }
+      if (isVMExitCode(res.error())) {
+        ABORT(static_cast<VMExitCode>(res.error().value()));
+      }
+      ABORT(default_error);
+    }
+    return outcome::success();
+  }
+
+  /**
+   * Aborts execution if res has error
+   * @tparam T - result type
+   * @param res - result to check
+   * @param default_error - default VMExitCode to abort with.
+   * @return If res has no error, res returned. Otherwise if res.error()
+   * is VMAbortExitCode or VMFatal, the res.error() returned, else
+   * if res.error()is VMExitCode, the aborted res.error()returned, else
+   * aborted default_error returned
+   */
+  template <typename T>
+  outcome::result<T> requireNoErrorAssign(outcome::result<T> &&res,
+                                          const VMExitCode &default_error) {
+    if (res.has_error()) {
+      if (isFatal(res.error()) || isAbortExitCode(res.error())) {
+        return res.error();
+      }
+      if (isVMExitCode(res.error())) {
+        ABORT(static_cast<VMExitCode>(res.error().value()));
+      }
+      ABORT(default_error);
+    }
+    return std::move(res);
+  }
+
+  /**
+   * Aborts execution if res is not success
+   * @tparam T - result type
+   * @param res - result to check
+   * @return If res has no error, success() returned. Otherwise VMAbortExitCode
+   * returned
+   */
+  template <typename T>
+  outcome::result<void> requireSuccess(const outcome::result<T> &res) {
+    if (res.has_error()) {
+      ABORT(res.error().value());
```

Comment:
Must check `asExitCode` before `ABORT` cast.
Any other error is returned as is.

Reply:
But spec actors aborts it anyway.
`if !e.IsSuccess() {
		rt.Abortf(e, msg, args...)
}`

Reply:
`Abortf` accepts `ExitCode`, not `error`

Reply:
Done